client/trino-client/src/main/java/io/trino/client/ResultRows.java

```diff
@@ -38,11 +43,56 @@ public Iterator<List<Object>> iterator()
         {
             return emptyIterator();
         }
+
+        @Override
+        public String toString()
+        {
+            return "EmptyResultRows{}";
+        }
     };
 
     static ResultRows wrapList(List<List<Object>> values)
     {
-        return values::iterator;
+        return new ResultRows() {
+            @Override
+            public void close() {}
+
+            @Override
+            public Iterator<List<Object>> iterator()
+            {
+                return values.iterator();
+            }
+
+            @Override
+            public String toString()
+            {
+                return "ResultRows{values=" + values + "}";
+            }
+        };
+    }
+
+    static ResultRows wrapIterator(CloseableIterator<List<Object>> iterator)
```

Comment:
Again we are pretending to have iterable while we can acquire iterator only once

Reply:
This should probably take iterator factory as input. And track iterators created with it. And close all on close

Reply:
Iterable is convenient from the consumer API point of view:

```
            ResultRows resultRows = clientV1.currentRows();
            if (!resultRows.isNull()) {
                for (List<Object> row : resultRows) {
                    rows++;
                }
            }
```

Reply:
I've added a safe guard so that iterator is acquired once from the ResultRows.